### ISSUE TITLE
Support DTO validation

### ISSUE THREAD

**Issue by ejball** [FacilityApi/FacilityCSharp#23]
We should have a standard way to validate DTO properties, e.g. via a `Validate` method that returns a `ServiceResult`, perhaps. To start with, it would respect the [`required` attribute](https://github.com/FacilityApi/Facility/issues/24), but it could potentially support more advanced validation in the future.

**Comment by ejball** [comment]
I think I'd like to add this method to `ServiceDto`:

```csharp
public virtual IEnumerable<string> GetValidationErrors() => Array.Empty<string>();
```

The code that generates each DTO could generate an override for this method that returns all of the errors associated with the DTO. It would respect the `required` attribute, and recursively validate any descendant DTOs.

**Comment by jzbrooks** [comment]
So the HTTP mapping generator would generate something like:

```csharp
new HttpMethodMapping<T, U>.Builder {    
    ValidateRequest = request => {
        var errors = request.Validate();
        if (errors.Count > 0) {
            return ServiceResult.Failure(ServiceErrors.CreateInvalidRequest(string.Join("\n", errors));
        }
        return ServiceResult.Success()
    }
}
```

Would that render `ServiceErrors.CreateRequestFieldRequired` obsolete?
Are there ever any cases where validation may result in a result other than the obvious invalid request?

**Comment by ejball** [comment]
Validation could happen at a higher level so that the same code doesn't need to be generated for every method.

`ServiceErrors.CreateRequestFieldRequired` could still be used when fields are conditionally required.

I've been thinking `ServiceError` should have an `errors` or `innerErrors` field. Or maybe `messages`? I hadn't thought to concatenate the messages, but maybe that wouldn't be so bad.

I'm not sure what you're asking about other transport layers.

**Comment by jzbrooks** [comment]
I'm curious if there are other situations where there may be different versions of an invalid request. In the example code with the HTTP mapper, every invalid request is lumped into a single type of response, namely a service error with the `InvalidRequest` code specified. Are there examples beyond HTTP that have some semantic notion of differences in invalid request scenarios? Perhaps that isn't something Facility is concerned about. I haven't looked into protocols like gRPC in detail to know.

To put it another way, your initial suggestion of a `Validate` method that returned a `ServiceResult` would allow each validation call to specify _how_ it failed validation. I'm not sure if that's important or not.

**Comment by ejball** [comment]
It is possible to define custom error codes; I think that gets the closest to what you are asking about. I haven't written any APIs that use that feature.

**Comment by jzbrooks** [comment]
It looks like there's already a pretty similar approach on [Facility.Definition.ServiceElementInfo.cs](https://github.com/FacilityApi/Facility/blob/master/src/Facility.Definition/ServiceElementInfo.cs) for FSD parsing.

Would the recursive function obviate the need for [the validation already done in code generation](https://github.com/FacilityApi/FacilityCSharp/blob/79c7ecfc323c37afdc8b4b6d987f369ce3c1086a/src/Facility.CodeGen.CSharp/CSharpGenerator.cs#L470-L499), or is that code generation responsible for HTTP-specific concerns and the recursive validation logic responsible for fsd-defined validation (like `[required]`)?

**Comment by ejball** [comment]
The latter.

**Comment by ejball** [comment]
I worked on this in PR #29. I decided to go with a single error message; reporting all validation errors just didn't seem worth the effort.

I settled on a `bool Validate(out string? errorMessage)` method for `ServiceDto`. It seemed the cleanest and clearest of the APIs I considered.
